doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,258 @@
+                 +----+----+ |
+                             |
+   +----+--------+           V
+   |cGRE| osdmap +-+       +----+--------+
+   +----+--------+ |       |cBLU| osdmap +-+
+                   |       +----+--------+ |
+                   V                       |
+                 +----+----+               V
+                 |cGRE| PG +-+           +----+--------+
+                 +----+----+ |           |cBLU| active +-+
+                             |           +----+--------+ |
+                             V                           |
+                           +----+--------+               V
+                           |cBLU| osdmap +-+           +----+----------+
+                           +----+--------+ |           |cBLU| recovery +-+
+                                           |           +----+----------+ |
+                                           V                             |
+                                         +----+--------+                 V
+                                         |cBLU| active +-+             +----+---------+
+                                         +----+--------+ |             |cBLU| obj ctx +-+
+                                                         |             +----+---------+ |
+                                                         V                              |
+                                                       +----+----------+                V
+                                                       |cBLU| recovery +-+             +----+---------+
+                                                       +----+----------+ |             |cBLU| process |
+                                                                         |             +----+---------+
+                                                                         V
+                                                                       +----+---------+
+                                                                       |cBLU| obj ctx +-+
+                                                                       +----+---------+ |
+                                                                                        |
+                                                                                        V
+                                                                                       +----+---------+
+                                                                                       |cBLU| process |
+                                                                                       +----+---------+
+
+In the diagram above, two requests from the same client are being served in parallel. They share the same
+pipeline. But before one transits to the next stage, it have to wait until the previous one leaves that stage.
+And please note, the last "process" stage is a per-PG stage. So only one request is allowed to stay in that
+stage at any give point of time.
```

Comment:
IIUC the major disadvantage of the current implementation is that we don't allow more than one request to be processed at the same time in the same PG. And we'd probably relax this by allowing restricted concurrency in a PG *and at the same time, make sure their replies are still ordered by client*. Specifically, the restricted concurrency means:
1. To allow read/write to different rados objects in the same PG to be concurrent;
2. To allow multiple reads to the same object to be concurrent if there are no interleaved writes;
3. To allow to reorder reads ahead of writes in the same object if they are not dependent;
4. We don't need to consider to optimize with concurrent/reordered writes in the same object, because we've already implemented striping at the level of rbd;

Is that precise?

Reply:
> We don't need to consider to optimize with concurrent/reordered writes in the same object, because we've already implemented striping at the level of rbd;

just a comment, in the use case of using object for storing indexes of application object, the concurrent/reordered writes in the same object might help, if the rados application uses in-place writes to update the index stored in a certain object, and the application updates the index when updating, creating or removing the indexed object.

after a second thought, i still think the concurrent read would benefit the use case of "index stored in object". for instance, rgw stores the index of the stored objects in a bucket, and the object and associated omap in that bucket would be a bottleneck of the read workload if we cannot offer good throughput for read ops targeting at a single object.

Reply:
See above for a partial but more detailed discussion of what would be involved.